uPy/html.py

```diff
@@ -0,0 +1,255 @@
+
+        #form tag: beginning of form
+        if(tag_internals[0]=="form"):
+
+            # Create a new dictionary to store info
+            tagDict = {}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("tagDict[tag_type]= form")
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key] = val
+                #print("\ttagDict[{0}]={1}".format(key,val))
+            
+            # other tags will be inside forms
+            tagDict["inside"] = []
+            form = tagDict
+            formOpen = True
+
+        # /form tag : end of form
+        elif(tag_internals[0]=="/form"):
+            #print("\ttag_type=/form")
+            formOpen=False #close out the form
+
+
+        # a tag: indicates hyperlinked text
+        elif(tag_internals[0]=="a" or tag_internals[0]=="input"):
+            # Create a new dictionary to store info
+            tagDict={}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("\ttagDict[\"tag_type\"]={0}".format(tag_internals[0]))
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key]=val
+                #print("\ttagDict[\"{0}\"]={1}".format(key,val))
+            
+            #append this tag into the form
+            if(formOpen):
+                    inside_list=form["inside"]   #get
+                    inside_list.append(tagDict)  #append
+                    form["inside"]=inside_list   #put back
+
+    return form
+
+
+# TOP LEVEL FUNCTION
+# Input: HTML splashpage as bytes
+# Returns a list of dictionaries. Each dictionary is a form:    
+#   Pick a form by indexing into the list. Each form is a dictionary.   
+#   Dictionary key value pairs are keys and values within that tag line:
+#        ex from html: name="formname" method="POST"
+#        becomes:   dict["name"] -> "formname"
+#                   dict["method"] -> "POST"
+def get_forms(html):
+    
+    html = html.decode('utf-8') # int to str
+
+    # each entry in list is "<form ...> <...> </form>" 
+    forms = get_tags(html, "form")
+    if(len(forms) == 0):
+        return []
+    elif(len(forms) == 1): #caused if <form...> but bad html received no </form>
+        forms = []      #get rid of previous bad entry
+        ret_val = find_complete_string(html, "<form", "</div", 0) #WORKAROUND: try going until the end of the next div
+        if(ret_val[0] == ""):
+            return []   #gah lee you've just hit the worst written splashpage ever. just give up already
+        forms.append(ret_val[0])
+    
+    #debugging
+    print("All forms as complete strings:")
+    i=0
+    while i < len(forms):
```

Comment:
is this print helpful, if so, it should be a for loop:

`for i, form in enumerate(forms):
     print("...".format(i, form))`